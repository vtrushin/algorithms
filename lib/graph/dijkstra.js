export default (graph, start, end) => {
	[start, end].forEach(vertex => {
		if (!graph[vertex]) {
			throw new Error(`Incorrect vertex name: "${vertex}". Must be one of "${Object.keys(graph).join(', ')}"`);
		}
	});

	const distances = {};
	const unvisited = new Set();
	const measuredUnvisited = new Set();
	const transitions = {};

	distances[start] = 0;
	measuredUnvisited.add(start);

	Object.keys(graph).forEach(vertex => {
		if (vertex !== start) {
			distances[vertex] = Infinity;
			unvisited.add(vertex);
		}
	});

	while (unvisited.size) {
		let selected;

		let minVertexDistance = Infinity;
		for (const vertex of measuredUnvisited) {
			if (distances[vertex] < minVertexDistance) {
				minVertexDistance = distances[vertex];
				selected = vertex;
			}
		}

		if (!selected || selected === end) {
			break;
		}

		unvisited.delete(selected);
		measuredUnvisited.delete(selected);

		const neighbors = graph[selected];

		for (const neighbor in neighbors) {
			if (neighbors.hasOwnProperty(neighbor)) {
				const distanceToNeighbor = neighbors[neighbor];

				if (!unvisited.has(neighbor)) {
					continue;
				}

				const newVertexDistance = distances[selected] + distanceToNeighbor;

				if (newVertexDistance < distances[neighbor]) {
					distances[neighbor] = newVertexDistance;
					transitions[neighbor] = selected;
					measuredUnvisited.add(neighbor);
				}
			}
		}
	}

	const path = [];

	let from = transitions[end];

	if (from) {
		path.push(end);
	}

	while (from) {
		path.push(from);

		if (from !== start) {
			from = transitions[from];
		} else {
			break;
		}
	}

	path.reverse();

	return path;
};
