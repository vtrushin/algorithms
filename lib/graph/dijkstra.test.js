import dijkstra from './dijkstra';

describe('Graph 1', () => {
	const graph = {
		A: { B: 7, C: 2 },
		B: { A: 7, D: 3, E: 2 },
		C: { A: 2, D: 3, E: 10 },
		D: { B: 1.5, C: 6, F: 8 },
		E: { B: 4, C: 10, F: 4 },
		F: { D: 8, E: 4 },
		G: { C: 1000 }
	};

	it('Short path', () => {
		expect(
			dijkstra(graph, 'A', 'D')
		).toEqual(
			['A', 'C', 'D']
		)
	});

	it('Unreachable vertex', () => {
		expect(
			dijkstra(graph, 'A', 'G')
		).toEqual(
			[]
		)
	});

	it('Incorrect vertex', () => {
		expect(() => {
			dijkstra(graph, 'A', 'INCORRECT')
		}).toThrowErrorMatchingSnapshot()
	});
});
