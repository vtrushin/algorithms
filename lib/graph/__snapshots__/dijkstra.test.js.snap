// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph 1 Incorrect vertex 1`] = `"Incorrect vertex name: \\"INCORRECT\\". Must be one of \\"A, B, C, D, E, F, G\\""`;
